src/coreclr/src/jit/morph.cpp

```diff
@@ -7298,306 +7283,724 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
+    }
+
+    GenTree* retAddrSlot           = gtNewOperNode(GT_ADDR, TYP_I_IMPL, gtNewLclvNode(lvaRetAddrVar, TYP_I_IMPL));
+    callDispatcherNode->gtCallArgs = gtPrependNewCallArg(retAddrSlot, callDispatcherNode->gtCallArgs);
+
+    if (origCall->gtType == TYP_VOID)
+    {
+        return callDispatcherNode;
+    }
+
+    assert(retVal != nullptr);
+    GenTree* comma = gtNewOperNode(GT_COMMA, origCall->TypeGet(), callDispatcherNode, retVal);
+    // The JIT seems to want to CSE this comma and messes up multi-reg ret
+    // values in the process. Just avoid CSE'ing this tree entirely in that
+    // case.
+    if (origCall->HasMultiRegRetVal())
+    {
+        comma->gtFlags |= GTF_DONT_CSE;
+    }
+
+    return comma;
+}
 
-        if (vtabOffsOfIndirection != CORINFO_VIRTUALCALL_NO_CHUNK)
+GenTree* Compiler::getMethodPointerTree(CORINFO_RESOLVED_TOKEN* pResolvedToken, CORINFO_CALL_INFO* pCallInfo)
+{
+    switch (pCallInfo->kind)
+    {
+        case CORINFO_CALL:
+            return new (this, GT_FTN_ADDR) GenTreeFptrVal(TYP_I_IMPL, pCallInfo->hMethod);
+        case CORINFO_CALL_CODE_POINTER:
+            return getLookupTree(pResolvedToken, &pCallInfo->codePointerLookup, GTF_ICON_FTN_ADDR, pCallInfo->hMethod);
+        default:
+            noway_assert(!"unknown call kind");
+            return nullptr;
+    }
+}
+
+GenTree* Compiler::getLookupTree(CORINFO_RESOLVED_TOKEN* pResolvedToken,
+                                 CORINFO_LOOKUP*         pLookup,
+                                 unsigned                handleFlags,
+                                 void*                   compileTimeHandle)
+{
+    if (!pLookup->lookupKind.needsRuntimeLookup)
+    {
+        // No runtime lookup is required.
+        // Access is direct or memory-indirect (of a fixed address) reference
+
+        CORINFO_GENERIC_HANDLE handle       = nullptr;
+        void*                  pIndirection = nullptr;
+        assert(pLookup->constLookup.accessType != IAT_PPVALUE && pLookup->constLookup.accessType != IAT_RELPVALUE);
+
+        if (pLookup->constLookup.accessType == IAT_VALUE)
         {
-            add = gtNewOperNode(GT_ADD, TYP_I_IMPL, vtbl, gtNewIconNode(vtabOffsOfIndirection, TYP_I_IMPL));
+            handle = pLookup->constLookup.handle;
+        }
+        else if (pLookup->constLookup.accessType == IAT_PVALUE)
+        {
+            pIndirection = pLookup->constLookup.addr;
+        }
 
-            GenTree* indOffTree = nullptr;
+        return gtNewIconEmbHndNode(handle, pIndirection, handleFlags, compileTimeHandle);
+    }
 
-            if (isRelative)
-            {
-                indOffTree = impCloneExpr(add, &add, NO_CLASS_HANDLE, (unsigned)CHECK_SPILL_ALL,
-                                          nullptr DEBUGARG("virtual table call"));
-            }
+    return getRuntimeLookupTree(pResolvedToken, pLookup, compileTimeHandle);
+}
+
+GenTree* Compiler::getRuntimeLookupTree(CORINFO_RESOLVED_TOKEN* pResolvedToken,
```

Comment:
Missing function header comments.

Reply:
These functions do not really work anyway, they are not faithful to their importer counterparts. It would be more wise to spend time to unify the runtime lookup generation in both phases instead (eg. through `GT_RUNTIMELOOKUP` node expanded later and used by both import and morph).

Reply:
Could you add a `TODO` comment?

Reply:
I hope to experiment with this soonish.

Reply:
> These functions do not really work anyway

Not sure how to parse this -- is there still work to do here? Experiments to try?

Reply:
In the current state these functions work correctly but in some cases they call runtime lookup helpers instead of generating inline code that requires introducing flow, which is what the importer counterparts do. See the TODO-CQ comment below.